File: hw/elbay/u3/hw25TemplateMethod/hw25TemplateMethod/python/model/meal_preparation.py
from abc import ABC, abstractmethod


class MealPreparation(ABC):
    def prepare_meal(self):
        self.gather_ingredients()
        self.cook()
        self.plate()
        if self.wants_garnish():
            self.add_garnish()

    def gather_ingredients(self):
        print("Gathering ingredients")

    @abstractmethod
    def cook(self):
        pass

    def plate(self):
        print("Plating the dish")

    @abstractmethod
    def add_garnish(self):
        pass

    def wants_garnish(self):
        return True

File: hw/elbay/u3/hw25TemplateMethod/hw25TemplateMethod/js/model/MealPreparation.js
class MealPreparation {
    prepareMeal() {
        this.gatherIngredients();
        this.cook();
        this.plate();
        if (this.wantsGarnish()) {
            this.addGarnish();
        }
    }

    gatherIngredients() {
        console.log("Gathering ingredients");
    }

    cook() {
        throw new Error("cook() must be implemented");
    }

    plate() {
        console.log("Plating the dish");
    }

    addGarnish() {
        throw new Error("addGarnish() must be implemented");
    }

    wantsGarnish() {
        return true;
    }
}

module.exports = MealPreparation;
